File: src/Runtime/Sentence/SentenceVariableAssign.cs
/*
	PeakScript: https://github.com/Jenocn/PeakScript
	By Jenocn: https://jenocn.github.io
*/

namespace peak.interpreter {

	class SentenceVariableAssign : Sentence {
		private SentenceExpression _variableExpression = null;
		private SentenceExpression _expression = null;
		public SentenceVariableAssign(SentenceExpression variableExpression, SentenceExpression expression) {
			_variableExpression = variableExpression;
			_expression = expression;
		}
		public override ExecuteResult Execute(Space space) {
			if (_variableExpression.GetExpressionType() != ExpressionType.Variable) {
				ErrorLogger.LogRuntimeError(ErrorRuntimeCode.VariableAssign, "The result of expression isn't a variable!");
				return ExecuteResult.Failed;
			}
			if (!IsSuccess(_variableExpression.Execute(space))) {
				ErrorLogger.LogRuntimeError(ErrorRuntimeCode.VariableAssign, "The expression execute failed!");
				return ExecuteResult.Failed;
			}
			var variable = (_variableExpression as SentenceExpressionVariable).variable;
			if (!IsSuccess(_expression.Execute(space))) {
				ErrorLogger.LogRuntimeError(ErrorRuntimeCode.VariableAssign, "The expression execute failed!");
				return ExecuteResult.Failed;
			}
			if (!variable.SetValue(_expression.value)) {
				ErrorLogger.LogRuntimeError(ErrorRuntimeCode.VariableAssign, "The expression execute failed!");
				return ExecuteResult.Failed;
			}
			return ExecuteResult.Successed;
		}

	}
} // namespace peak.interpreter

File: src/Runtime/Sentence/sentence_variable_assign.cpp

#include "sentence_variable_assign.h"
#include "runtime/variable.h"
#include "sentence_expression.h"
#include "sentence_expression_variable.h"

using namespace peak;

SentenceVariableAssign::SentenceVariableAssign(std::shared_ptr<SentenceExpression> variableExpression, std::shared_ptr<SentenceExpression> expression)
	: _variableExpression(variableExpression), _expression(expression) {
}

ExecuteResult SentenceVariableAssign::Execute(std::shared_ptr<Space> space) {
	if (_variableExpression->GetExpressionType() != ExpressionType::Variable) {
		ErrorLogger::LogRuntimeError(ErrorRuntimeCode::VariableAssign, "The result of expression isn't a variable!");
		return ExecuteResult::Failed;
	}
	if (!IsSuccess(_variableExpression->Execute(space))) {
		ErrorLogger::LogRuntimeError(ErrorRuntimeCode::VariableAssign, "The expression execute failed!");
		return ExecuteResult::Failed;
	}
	auto variable = std::static_pointer_cast<SentenceExpressionVariable>(_variableExpression)->GetVariable();
	if (!IsSuccess(_expression->Execute(space))) {
		ErrorLogger::LogRuntimeError(ErrorRuntimeCode::VariableAssign, "The expression execute failed!");
		return ExecuteResult::Failed;
	}
	if (!variable->SetValue(_expression->GetValue())) {
		ErrorLogger::LogRuntimeError(ErrorRuntimeCode::VariableAssign, "The expression execute failed!");
		return ExecuteResult::Failed;
	}
	return ExecuteResult::Successed;
}